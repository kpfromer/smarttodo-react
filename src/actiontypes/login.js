export const INVALIDATE_LOGIN = 'login/INVALIDATE_LOGIN';
export const REQUEST_LOGIN = 'login/REQUEST_LOGIN';
export const RECEIVE_LOGIN = 'login/RECEIVE_LOGIN';
export const FAILURE_LOGIN = 'login/FAILURE_LOGIN';