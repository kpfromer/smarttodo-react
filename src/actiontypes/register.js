
export const REQUEST_REGISTER = 'register/REQUEST_REGISTER';
export const RECEIVE_REGISTER = 'register/RECEIVE_REGISTER';
export const FAILURE_REGISTER = 'register/FAILURE_REGISTER';