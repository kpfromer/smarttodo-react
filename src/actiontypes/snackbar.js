export const CLOSE_SNACKBAR = 'snackbar/CLOSE_SNACKBAR';
export const CHANGE_SNACKBAR = 'snackbar/CHANGE_SNACKBAR';