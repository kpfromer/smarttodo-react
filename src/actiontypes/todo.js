export const REQUEST_TODOS = 'todo/REQUEST_TODOS';
export const REQUEST_FAILURE_TODOS = 'todo/REQUEST_FAILURE_TODOS';
export const RECIEVE_TODOS = 'todo/RECIEVE_TODOS';

export const ADD_TODO = 'todo/ADD_TODO';
export const ADD_TODO_FAILURE = 'todo/ADD_TODO_FAILURE';
export const ADD_TODO_SUCCESS = 'todo/ADD_TODO_SUCCESS';

export const UPDATE_TODO = 'todo/UPDATE_TODO';
export const UPDATE_TODO_SUCCESS = 'todo/UPDATE_TODO_SUCCESS';
export const UPDATE_TODO_FAILURE = 'todo/UPDATE_TODO_FAILURE';

export const REMOVE_TODO = 'todo/REMOVE_TODO';
export const REMOVE_TODO_SUCCESS = 'todo/REMOVE_TODO_SUCCESS';
export const REMOVE_TODO_FAILURE = 'todo/REMOVE_TODO_FAILURE';