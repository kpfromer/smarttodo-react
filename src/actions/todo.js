import * as TodoActionTypes from '../actiontypes/todo';
import uuid from 'uuid/v4';
import { CALL_API } from "../middleware/api/api";

export const fetchTodos = () => ({
  [CALL_API]: {
    types: [
      TodoActionTypes.REQUEST_TODOS,
      TodoActionTypes.RECIEVE_TODOS,
      TodoActionTypes.REQUEST_FAILURE_TODOS
    ],
    method: 'GET',
    endpoint: '/todo',
    minimumDelay: 1000
  }
});

export const addTodo = (description, completed) => {
  const todo = { description, completed };
  return {
    tempId: uuid(),
    todo,
    [CALL_API]: {
      types: [
        TodoActionTypes.ADD_TODO,
        TodoActionTypes.ADD_TODO_SUCCESS,
        TodoActionTypes.ADD_TODO_FAILURE
      ],
      method: 'POST',
      endpoint: '/todo',
      data: todo
    }
  }
};

export const updateTodo = (prevTodo, newTodo, id) => {

  const { id: unUsedId, ...newTodoWithoutId } = newTodo;

  return {
    id,
    todo: newTodo,
    revert: {
      todo: prevTodo
    },
    [CALL_API]: {
      types: [
        TodoActionTypes.UPDATE_TODO,
        TodoActionTypes.UPDATE_TODO_SUCCESS,
        TodoActionTypes.UPDATE_TODO_FAILURE
      ],
      method: 'PUT',
      endpoint: `/todo/${id}`,
      data: newTodoWithoutId
    }
  }
};

export const removeTodo = (todo, position, id) => ({
  id,
  revert: {
    todo,
    position
  },
  [CALL_API]: {
    types: [
      TodoActionTypes.REMOVE_TODO,
      TodoActionTypes.REMOVE_TODO_SUCCESS,
      TodoActionTypes.REMOVE_TODO_FAILURE
    ],
    method: 'DELETE',
    endpoint: `/todo/${id}`
  }
});
