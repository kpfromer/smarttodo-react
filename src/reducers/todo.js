import * as TodoActionTypes from '../actiontypes/todo';

function todos(
  state = {
    isFetching: false,
    didInvalidate: false,
    todos: []
  },
  action
) {
  switch (action.type) {
    case TodoActionTypes.REQUEST_TODOS:
      return {
        ...state,
        isFetching: true,
        didInvalidate: false
      };
    case TodoActionTypes.RECIEVE_TODOS:
      return {
        ...state,
        isFetching: false,
        didInvalidate: false,
        todos: action.response
      };
    case TodoActionTypes.ADD_TODO:
      // TODO: fix backend usage so it is not an array
      return {
        ...state,
        todos: [
          ...state.todos,
          {
            ...action.todo,
            id: action.tempId
          }
        ]
      };
    case TodoActionTypes.ADD_TODO_SUCCESS:
      const tempTodoIndex = state.todos.findIndex(todo => todo.id === action.tempId);
      const tempTodo = state.todos[tempTodoIndex];
      if (tempTodoIndex < 0) {
        throw new Error('Todo not found!');
      }
      // TODO: same as ADD_TODO - fix backend usage so it is not an array
      return {
        ...state,
        todos: [
          ...state.todos.slice(0, tempTodoIndex),
          {
            ...tempTodo,
            id: action.response[0]._id
          },
          ...state.todos.slice(tempTodoIndex + 1)
        ]
      };
    case TodoActionTypes.ADD_TODO_FAILURE:
      const tempTodoRemoveIndex = state.todos.findIndex(todo => todo.id === action.tempId); // TODO: extract
      return {
        ...state,
        todos: [
          ...state.todos.slice(0, tempTodoRemoveIndex),
          ...state.todos.slice(tempTodoRemoveIndex + 1)
        ]
      };
    case TodoActionTypes.REMOVE_TODO:
      return {
        ...state,
        todos: state.todos.filter(todo => todo.id !== action.id)
      };
    case TodoActionTypes.REMOVE_TODO_FAILURE:
      const { todo: oldTodo, position: oldTodoPosition } = action.revert;
      return {
        ...state,
        todos: [
          ...state.todos.slice(0, oldTodoPosition),
          oldTodo,
          ...state.todos.slice(oldTodoPosition)
        ]
      };
    case TodoActionTypes.UPDATE_TODO:
      return {
        ...state,
        todos: state.todos.map(todo => todo.id === action.id ? action.todo : todo)
      };
    case TodoActionTypes.UPDATE_TODO_FAILURE:
      return {
        ...state,
        todos: state.todos.map(todo => todo.id === action.id ? action.revert.todo : todo)
      };
    default:
      return state;
  }
}

export default todos;