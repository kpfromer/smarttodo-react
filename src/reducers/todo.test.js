import TodoReducer from './todo';
import * as TodoActionTypes from "../actiontypes/todo";

// TODO: refactor test to follow nature flow
// TODO: Example for above request -> error, request -> success
describe('todo reducer', () => {
  let currentState;

  beforeEach(() => {
    currentState = {
      isFetching: false,
      didInvalidate: false,
      todos: [
        {
          id: '1',
          description: 'other',
          completed: true
        },
        {
          id: '2',
          description: 'todos',
          completed: false
        }
      ]
    };
  });

  it('should display spinner on todos request', () => {
    expect(TodoReducer(
      currentState,
      {
        type: TodoActionTypes.REQUEST_TODOS
      }
    )).toMatchSnapshot();
  });

  it('should replace todos from api response', () => {
    const completelyOtherTodos = [
      {
        id: '13242352',
        description: 'math science, chem',
        completed: false
      },
      {
        id: '2dfasdfasdf',
        description: 'different',
        completed: true
      },
      {
        id: '12345',
        description: 'ss test',
        completed: false
      }
    ];
    expect(TodoReducer(
      currentState,
      {
        type: TodoActionTypes.RECIEVE_TODOS,
        response: completelyOtherTodos
      }
    )).toMatchSnapshot();
  });

  describe('ADD_TODO', () => {
    let tempId;

    beforeEach(() => {
      tempId = 'a-temp-id';
      currentState = {
        ...currentState,
        todos: [
          {
            id: '13242352',
            description: 'math science, chem',
            completed: false
          },
          {
            id: '2dfasdfasdf',
            description: 'different',
            completed: true
          },
          {
            id: tempId,
            description: 'old todo',
            completed: false
          }
        ]
      };
    });

    it('should add a todo', () => {
      const newState = TodoReducer(
        currentState,
        {
          type: TodoActionTypes.ADD_TODO,
          todo: {
            description: 'New Task',
            completed: true
          },
          tempId: 'a-temp-id-if-needed-to-remove-todo'
        }
      );

      expect(newState).toMatchSnapshot();
    });

    it('should remove todo if the add todo api call fails', () => {
      expect(TodoReducer(
        currentState,
        {
          type: TodoActionTypes.ADD_TODO_FAILURE,
          tempId
        }
      )).toMatchSnapshot();
    });
    it('should replace the added todo with the real id from api response', () => {
      expect(TodoReducer(
        currentState,
        {
          type: TodoActionTypes.ADD_TODO_SUCCESS,
          tempId,
          response: [
            {
              _id: 'api-database-id!'
            }
          ]
        }
      )).toMatchSnapshot();
    });
  });

  describe('REMOVE_TODO', () => {
    it('should remove a todo', () => {
      const newState = TodoReducer(
        currentState,
        {
          type: TodoActionTypes.REMOVE_TODO,
          id: '1'
        }
      );

      expect(newState).toMatchSnapshot();
    });
    it('should revert to original todo if remove todo api call fails', () => {
      const todo = {
        id: '1',
        description: 'reverted todo',
        completed: true
      };
      const newState = TodoReducer(
        currentState,
        {
          type: TodoActionTypes.REMOVE_TODO_FAILURE,
          id: '1',
          revert: {
            todo,
            position: 0
          }
        }
      );

      expect(newState).toMatchSnapshot();
    });
  });

  describe('UPDATE_TODO', () => {
    it('should update a todo', () => {
      const todo = {
        id: '1',
        description: 'updated description',
        completed: false
      };
      const newState = TodoReducer(
        currentState,
        {
          type: TodoActionTypes.UPDATE_TODO,
          id: todo.id,
          todo
        }
      );

      expect(newState).toMatchSnapshot();
    });
    it('should revert to original todo if update todo api call fails', () => {
      const todo = {
        id: '1',
        description: 'reverted',
        completed: true
      };
      const newState = TodoReducer(
        currentState,
        {
          type: TodoActionTypes.UPDATE_TODO_FAILURE,
          id: todo.id,
          revert: {
            todo
          }
        }
      );

      expect(newState).toMatchSnapshot();
    });
  });
});