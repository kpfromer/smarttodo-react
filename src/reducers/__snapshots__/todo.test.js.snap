// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo reducer ADD_TODO should add a todo 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": false,
      "description": "math science, chem",
      "id": "13242352",
    },
    Object {
      "completed": true,
      "description": "different",
      "id": "2dfasdfasdf",
    },
    Object {
      "completed": false,
      "description": "old todo",
      "id": "a-temp-id",
    },
    Object {
      "completed": true,
      "description": "New Task",
      "id": "a-temp-id-if-needed-to-remove-todo",
    },
  ],
}
`;

exports[`todo reducer ADD_TODO should remove todo if the add todo api call fails 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": false,
      "description": "math science, chem",
      "id": "13242352",
    },
    Object {
      "completed": true,
      "description": "different",
      "id": "2dfasdfasdf",
    },
  ],
}
`;

exports[`todo reducer ADD_TODO should replace the added todo with the real id from api response 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": false,
      "description": "math science, chem",
      "id": "13242352",
    },
    Object {
      "completed": true,
      "description": "different",
      "id": "2dfasdfasdf",
    },
    Object {
      "completed": false,
      "description": "old todo",
      "id": "api-database-id!",
    },
  ],
}
`;

exports[`todo reducer REMOVE_TODO should remove a todo 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": false,
      "description": "todos",
      "id": "2",
    },
  ],
}
`;

exports[`todo reducer REMOVE_TODO should revert to original todo if remove todo api call fails 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": true,
      "description": "reverted todo",
      "id": "1",
    },
    Object {
      "completed": true,
      "description": "other",
      "id": "1",
    },
    Object {
      "completed": false,
      "description": "todos",
      "id": "2",
    },
  ],
}
`;

exports[`todo reducer UPDATE_TODO should revert to original todo if update todo api call fails 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": true,
      "description": "reverted",
      "id": "1",
    },
    Object {
      "completed": false,
      "description": "todos",
      "id": "2",
    },
  ],
}
`;

exports[`todo reducer UPDATE_TODO should update a todo 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": false,
      "description": "updated description",
      "id": "1",
    },
    Object {
      "completed": false,
      "description": "todos",
      "id": "2",
    },
  ],
}
`;

exports[`todo reducer should display spinner on todos request 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": true,
  "todos": Array [
    Object {
      "completed": true,
      "description": "other",
      "id": "1",
    },
    Object {
      "completed": false,
      "description": "todos",
      "id": "2",
    },
  ],
}
`;

exports[`todo reducer should replace todos from api response 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "todos": Array [
    Object {
      "completed": false,
      "description": "math science, chem",
      "id": "13242352",
    },
    Object {
      "completed": true,
      "description": "different",
      "id": "2dfasdfasdf",
    },
    Object {
      "completed": false,
      "description": "ss test",
      "id": "12345",
    },
  ],
}
`;
